### ISSUE TITLE
GitJournal overwrites changes written in files on the desktop and pushed to Git

### ISSUE THREAD

**Issue by ovichiro** [GitJournal/GitJournal#241]
## Environment

**App version:** 
1.68.0

**Device information:**
 - OS name and version: Android 8.0
 - Manufacturer: Samsung
 - Model: A8

## Description

With either auto sync or manual sync, the app will overwrite changes pushed from the desktop.

**Expected behavior:**

The app should either lock the notes until the sync has finished in the case of auto-sync, or simply duplicate the file, one copy for desktop changes, one for mobile.
Or even attempt to merge.

**Current behavior:**

See title.

## Steps to reproduce

Enable auto sync in GitJournal.  
From the app, press the back button to exit to home screen.  

My desktop is synced to the same git repo.  
Add text to some files from the desktop text editor. One of the files is named 'test.md'.  
Save, commit, push.  

After some time, open GitJournal and edit test.md without waiting for the sync icon to finish.   
Press 'back' to close the note and stay on the note list screen.  
Git Journal will do a merge that overwrites the changes done on the desktop.  

### Manual sync case

Disable auto sync in GitJournal.  
Create a test file named "text.txt" on the desktop.  

Enter the contents:  

    Desktop test 1.
    Desktop test 2.

Git add, commit, push.  

Sync GitJournal so we have the file there.  

From the desktop add a line between the existing 2.  

    Desktop test 1.
    Desktop test 1.1.
    Desktop test 2.

Git add, commit, push.  

From GitJournal add a line between the existing 2, without sync.  

    Desktop test 1.
    Journal test 1.1.
    Desktop test 2.

Sync GitJournal by swiping down.  

Result: The line "Desktop test 1.1." will have been replaced by "Journal test 1.1."  

This simulates editing a note when no cell signal is available, in a subway, in the countryside etc.

**Comment by agittins** [comment]
I've replicated this on my setup. If you edit a "stale" note on the app, it will overwrite any existing change in the repo that applied to that same area of text.

Does the app do the same commit && pull --rebase && push cycle that is recommended for the desktop client?

This does appear to be limited to the case where it's the same "chunk" of text that's edited - if I make changes in areas that are separate (ie, that have an unchanged chunk of text for diff to lock to in between the changed bits) then it seems to work as expected.

**Comment by vHanda** [comment]
The app currently does a merge and not a rebase. It additionally does a merge by chosing the local version over the remote. It doesn't even attempt to merge the files at all.

I know, this is quite bad - this is something I definitely want to fix, and soon. Ideally there should be some smart merge, and configurable options on what to do when there is a conflict.

I'm going to rename this issue to reflect that a proper merge mechanism needs to be implemented.

This is easily going to take me 1-2 months, as I would like to solve it by improving my git library written in Dart, instead of using libgit2 for performing the merge.

With libgit2, each call goes via Flutter -> Android -> NDK -> Custom Lib -> libgit2. This is very cumbersome. 

(I can make configuring chosing local or remote changes very easily, I'll open another issue for it)

**Comment by ginkgomzd** [comment]
I can definitely work around this and recommend working on other more important features before taking on this rabbit hole. 

I do think the way VS Code does a GUI merge is the best example to follow, when you get to it. 

A high priority feature should be to just provide a warning to the user. The work around could be to offer to create a copy of the local version before loading the incoming changes.

**Comment by ovichiro** [comment]
The most honest fix is to detect the conflict and create 2 copies of the file, each with its own changes.   

> (I can make configuring chosing local or remote changes very easily, I'll open another issue for it)

In most cases, especially in long notes, I won't know which I want to keep. I won't remember. Probably both.  
With 2 copies at least I'm safe from losing data. It gives you time to implement other conflict resolutions and test them properly. Just my 2 cents.

**Comment by ginkgomzd** [comment]
@ovichiro there is no data loss. You just have to go into the git repo to retrieve the previous commit.
It's out of app, but being able to use git seems like a fair assumption for any user of this app.

**Comment by ovichiro** [comment]
@ginkgomzd You're right. But a notification is needed in this case, as you mentioned, so one would know there was a conflict. Otherwise, if I figure it out 2 weeks later, it might be cumbersome to dig through all the diffs.

**Comment by vHanda** [comment]
Could everyone interested in this please vote on this issue. I can accordingly prioritize it.

The simple options of -
1. Provide an option of which to choose local/remote
2. Duplicate the file silently

These both can be done right now. A more complex solution will require more time.

**Comment by ginkgomzd** [comment]
I'm ambivalent but I think saving the local changes to an obviously named new file is slightly better because I think you could manually merge the changes without leaving the app (basic copy-paste).

The other way, if you realize you want to merge the changes, you have to choose to keep the local changes, and then go into git to view the remote changes and decide if you want to restore them.

**Comment by MotaOcimar** [comment]
I think duplicating is a good choice.
Different names and a user warning will also help the user to figure this out and do something soon.
The [Evernote](https://help.evernote.com/hc/en-us/articles/209005237-How-does-sync-work-within-Evernote-) and the [Standard Notes](https://docs.standardnotes.org/specification/sync/#sync-completion) apps do something like that.

**Comment by agittins** [comment]
Oh god, please no! Duplicate notes are a key reason why I LEFT evernote! (they have a 5 year old bug where a patchy network connection results in dozens of duplicated notes). Once you create a duplicate file, how is the user (on a phone) expected to resolve that conflict and merge it back into the repo in a single file?

Between those two options, I'd *strongly* prefer (1) prompt the user. But I'd suggest not prompting to accept local vs remote, but prompt the user to tell them that it found a conflict, and take them to the file - with git's standard "merge fail" block that shows what each side has. That way both changesets are preserved in the visible file, and the user can tidy things up from there.

I am not all that experienced with git, but my point is that this is a problem that git has already "solved" as much as it can, so let's use git's solution to the problem, and just add notification/assistance on top of that, rather than creating a new method of resolving it.

**Comment by agittins** [comment]
I just got bitten by this again. @ginkgomzd you mentioned that there is no data loss, that it's in the git repo - but I can't see it there (I'm talking about the github repo and my desktop - maybe it's present locally on my phone?).

It looks to me like straight up data-loss - functionally, gitjournal has deleted a set of changes from github and from my desktop - this doesn't seem a good method.

Am I making it worse in how I sync my desktop? I'm doing:

```bash
git add --all
git commit -m "Autocommit from $(hostname)"
git pull --rebase
git push
```

Is that sensible, or is there something different I should do there? My reasoning is that if the rebase fails, I should have a merge conflict block sitting in the relevant file, with both the old and the new (or local and remote) changes, leaving it to me to sort out - which is perfect for me. It's visible (ie, it's right there in the file when I go to look at that data), it's complete (includes both conflicting blocks) and it's safe (no data is discarded).

Am I doing something dumb on the desktop end of it, or is this because of how gitjournal is pushing the changes?

**Comment by vHanda** [comment]
Sorry. I'm a very busy with personal stuff this week, but I've made a lot of progress in implementing this. I can give you a rough ETA of mid November.

**Comment by agittins** [comment]
That's totally cool! :-)

Is the plan to have conflicts appear as a merge conflict block within the file, just like git usually behaves? I'd really prefer that as a solution.

I have strong objections to duplicate files (they're invisible if you're using scripts/shortcuts/muscle-memory to edit notes), and to choosing local vs remote ("You have two children. Which one would you like me to destroy?"). A merge block I can address simply and at any time. Actually, I wouldn't mind if the merge conflict got comitted into the repo - it's text, not code, so doesn't matter if it's a bit broken-looking as long as everything's there.

In the interim I think I'll setup a cron job to sync my desktop regularly, and I'll take extra care to refresh my android before starting any edits.

Love your work!

**Comment by vHanda** [comment]
The main change will be trying to do a smart word based merge by default. Additionally, I'm going to be exposing the history of the file in an easily navigable option.

Regarding what to do when one cannot merge it, I don't know - for now, my plan is just to implement both options (copy a file, or show a dialog). I don't know which would make a better default.

**Comment by ginkgomzd** [comment]
@agittins that sucks you lost some changes.
I don't see anything wrong with your desktop sync script.

Since you are auto-committing from desktop in addition to the app, I don't see how you could have lost the data. There are cases where you can lose data in git, but that typically involves doing a force-push, which I don't see in your script.

When I say "repo": git is distributed version control... every copy has a repo. You inspect it by viewing the  `git log` and you can use `git show` to look at the changes in specific commits. Note that `git whatchanged` hides some commits, especially merge-commits.

**Comment by agittins** [comment]
If forced to choose between those two options, I'd definitely go for a dialog. Let the user see a "diff", and buttons to choose the action:
- Discard Local Changes
- Discard Remote Changes
- Create separate copy
- Merge differences

Where "merge differences" saves the diff chunks inline into the file and commits that. This would ideally result in a file that looks like the result of a failed/in-progress merge in git, or like the output of `diff -D`:

```
the number of planets are
<<<<<<< HEAD
nine
=======
eight
>>>>>>> branch-b
```


@ginkgomzd yes, you're right - I found the "missing" commits in my local history using `git reflog` - the commits were not in `git log`, and they were not visible on github (but they were probably in the history, just not visible in github's web UI).

My biggest worry is that this missing data was "invisible" - I had to go look for it to work out it was missing. If I hadn't noticed the data missing, I would never have found it. I was lucky it was my main inbox/todolist file.

**Comment by agittins** [comment]
So one can look at the equivalent of a "reflog" on github using the API (`curl -u username:token https://api.github.com/repos/username/reponame/events`) and I can see that the event from gitjournal was two commits, one is just the diff that applied the two-line edit I made, and the other was the auto-merge that reverted the entire repo back to the state that gitjournal last updated (which seems to have been about 10 days prior).

I had made a sweeping lint-style change to my entire notes setup in that time (fixing list formats) so there were 11,145 edits to 308 files. They were all silently rolled back. Well, merged out. None of this was visible in the github UI, or using `git log` on my local machine (only through git reflog or the github API). 

At first I thought this bug was just about the occasional risk of losing one's latest commit if there was a genuine merge conflict, but this appears much more serious, gitjournal is blindly pushing up the state of it's entire local repo, which might be of ANY age. Am I mis-reading the situation?

@vHanda would it be easy / possible / workable to change gitjournal to do a pull --rebase or at least just a fetch before it does anything else? It looks like right now, if you do an edit before it refreshes, it effectively reverts the entire repo back to the device's last fetch - which for a rarely-used device might be weeks, and it does it in a way that is invisible to everything short of `git reflog`.

**Comment by vHanda** [comment]
> 
> @vHanda would it be easy / possible / workable to change gitjournal to do a pull --rebase or at least just a fetch before it does anything else? It looks like right now, if you do an edit before it refreshes, it effectively reverts the entire repo back to the device's last fetch - which for a rarely-used device might be weeks, and it does it in a way that is invisible to everything short of `git reflog`.

For the last month or so I switched to - fetch + merge.

**Comment by ginkgomzd** [comment]
This makes me think that a simple improvement might be to have a stronger warning if a fetch wasn't successful recently. Could maybe keep things as is for listings, but add something very prominent if editing a file and a fetch hasn't succeeded since the app resumed.

**Comment by vHanda** [comment]
> This makes me think that a simple improvement might be to have a stronger warning if a fetch wasn't successful recently. Could maybe keep things as is for listings, but add something very prominent if editing a file and a fetch hasn't succeeded since the app resumed.

How about I make the sync icon (cloud) into something more prominent which indicates an error and clicking on it shows the last time a fetch worked and the current error?

**Comment by ginkgomzd** [comment]
> How about I make the sync icon (cloud) into something more prominent which indicates an error and clicking on it shows the last time a fetch worked and the current error?

I think I've been trained by the app to note the status of the sync icon, so more prominent doesn't feel like it would make a difference for me. My suggestion is really about having an alert on the file, view, when someone is about to make edits.

Just one user's opinion.

**Comment by vHanda** [comment]
Hello. Just wanted to give a mini update - I had to stop working on GitJournal for some health issues and now I'm slowly getting back into it. Hopefully I can tackle this soon.

**Comment by agittins** [comment]
> I had to stop working on GitJournal for some health issues and now I'm slowly getting back into it.

Sorry to hear that @vHanda, thanks for the update and I hope things are on the mend for you :-)

**Comment by MotaOcimar** [comment]
@agittins said:

> If forced to choose between those two options, I'd definitely go for a dialog. Let the user see a "diff", and buttons to choose the action:
> - Discard Local Changes
> - Discard Remote Changes
> - Create separate copy
> - Merge differences
> 
> Where "merge differences" saves the diff chunks inline into the file and commits that. This would ideally result in a file that looks like the result of a failed/in-progress merge in git, or like the output of `diff -D`:
> 
> ```
> the number of planets are
> <<<<<<< HEAD
> nine
> =======
> eight
> >>>>>>> branch-b
> ```
> 


Despite my previous comment, I have to agree with @agittins.  Git already does an excellent job of identifying and locating the conflict.  We may not need to innovate as much, just having the option "merge differences" will be a good choice for me (but the other suggested actions are also welcome).

**Comment by AlbertoFabbri93** [comment]
I think that sticking to the standard way git uses to deal with conflicts is the best idea. It feels natural since the notes are stored in a git repository and it puts the user in charge of deciding what to do, without having to guess.

**Comment by MotaOcimar** [comment]
Hi @vHanda!
Although I am new to Flutter, I want to help with this feature.
Have you any directions and advices for me and anyone who is trying to handle with merge on flutter?
Thanks!

**Comment by vHanda** [comment]
> Hi @vHanda!
> Although I am new to Dart & Flutter, I want to help with this feature.
> Have you any directions and advices for me and anyone who is trying to handle with merge on Dart?
> Thanks!

Hey MotaOcimar. The first step would be implementing a basic git merge in dart-git. I've already implemented `merge-base`, which is the most complicated part. You'll need at least a basic understand of the git internals. Please, let me know if you're still interested, I can provide more info.

Cheers

**Comment by MotaOcimar** [comment]
Oh, I have no knowledge of git internals, but I'm still interested.
It will probably take me a while to do something, since I will have to learn it, but it seems ok to me.
Sorry if I'm being naive trying to do something I don't know how to do yet ^^"
But I think it's worth a try

**Comment by vHanda** [comment]
> Oh, I have no knowledge of git internals, but I'm still interested.
> It will probably take me a while to do something, since I will have to learn it, but it seems ok to me.
> Sorry if I'm being naive trying to do something I don't know how to do yet ^^"
> But I think it's worth a try

Please don't apologize. All that matters is that you want to try and are willing to learn. How about you join the discord channel and we talk over there? It'll definitely be faster.

**Comment by vHanda** [comment]
Hello wonderful people.

I have news. If you look in Settings -> Experimental Features, there are now 2 new options "Dart only Git Implementation" and "Dart only merge implementation". These both shouldn't change the behavior of GitJournal in anyway, but they do stop using `libgit2` and use my own implementation of git in dart.

I would appreciate if you could test them out. They pass my tests, but you never know. The sooner I'm confident that dart-git works well, the sooner I can start using it and have a custom merge mechanism.

**Comment by craignicol** [comment]
I can see the options, but I can only enable "merge implementation", the toggle for "Git implementation" doesn't move.

**Comment by vHanda** [comment]
> I can see the options, but I can only enable "merge implementation", the toggle for "Git implementation" doesn't move.

Copy paste error. Fixed. I'll make a new release over the weekend.

**Comment by slayer** [comment]
Hi, thanks for your work.

I have a question, is an experimental  "Dart only merge implementation" just makes precedence local files over remote ?
As I see on my repo it just deletes  files, isn't?

**Comment by vHanda** [comment]
> Hi, thanks for your work.
> 
> I have a question, is an experimental "Dart only merge implementation" just makes precedence local files over remote ?

Yes. That's all for now. I was hoping to get feedback from users to know if they have encountered any problems, even though the code has tests.

After which, I can implement an actual improvement.

> As I see on my repo it just deletes files, isn't?

What do you mean by deletes files? This sounds scary.

**Comment by slayer** [comment]
> What do you mean by deletes files? This sounds scary.
I mean my remote file was zeroed

**Comment by vHanda** [comment]
@slayer : So if I understand correctly, after the merge the file was deleted instead of it choosing the remote file or local file?

**Comment by slayer** [comment]
> @slayer : So if I understand correctly, after the merge the file was deleted instead of it choosing the remote file or local file?

Yes, there was no choosing

**Comment by vHanda** [comment]
@slayer : I would love for you to share some of the info about your git repo. Here are the steps -

1. Take the two parents commits which were merged. This can be found by doing a `git cat-file commit <sha1Hash>`. You can look under `parent`.
2. For each of these commits, please look at the tree hash. This can also be found via `git cat-file`.
3. Use `git ls-tree <treeHash>` to list the contents of the folder. Please find the tree hash of the parent folder with the offending file, and give me its `ls-tree` output.
4. Please also provide the path to the folder.

If you would prefer to share this information via email feel free to email me - vhanda at gitjournal.io

**Comment by slayer** [comment]
Sent by email. 
BTW it happen again today

**Comment by mycrono** [comment]
After fighting with git for half a day I have decided to manually examine every merge because I don't trust any automatic merging anymore and also I don't trust `pull --rebase`.
Probably I am missing something obvious or I am just stupid.
Anyways, I wanted to share my current solution with you guys:

- GitJournal is set to separate branch "git-journal-shift"
- VS Code is set to branch "main"
- Whenever i want to sync or commit changes I execute the script smerge.sh (see bellow)
- if there is something to merge, the script will start the merge and then exit, so I can examine what exactly is changed
- after I commit the merge, I rerun the script smerge.sh to sync again

This complicated procedure is due to the fact that GitJournal pulls the changes and then overwrites them without checking for conflicts. Thus it creates a path of commits which tells Git that all conflicts have been resolved by GitJournal. I have found no way to tell Git that it should always produce a merge conflict if a file has been changed in both main and "git-journal-shift" reliably.

Here's smerge.sh

```bash
#!/usr/bin/env bash

# print executed commands
set -o xtrace

# fail fast
set -Eeuo pipefail

branch='git-journal-shift'
commit_message='m'

# abort if merge is ongoing
MERGE_INDICATOR_FILE="$(git rev-parse --git-dir)"/MERGE_HEAD
if test -f "$MERGE_INDICATOR_FILE"; then
    echo "*********************** merge is ongoing, commit first. ***********************"
    exit 1;
fi

# ref: https://stackoverflow.com/a/2658301
function evil_git_num_untracked_files {
  expr `git status --porcelain 2>/dev/null| grep "^??" | wc -l` 
}

# commit & pull & push
# git commit returns exit status 1 if nothing to commit so need to check first
if (( $(evil_git_num_untracked_files) > 0 )); then
  # there are untracked files
  git add -A
  git commit -a -m 'm'
fi
# if HEAD is dirty commit
git diff-index --quiet HEAD || git commit -a -m "$commit_message"
git pull
git push

# make merge visible
git fetch origin $branch:$branch
git merge --no-commit --no-ff $branch
git checkout --conflict merge .
if test -f "$MERGE_INDICATOR_FILE"; then
    echo "*********************** resolve merge and commit. ***********************"
    exit 1;
fi

# push branch
git branch --force $branch
git push origin $branch

# 
echo "*********************** ALL GOOD! **************************"
```

**Comment by mycrono** [comment]
I could not let it go and have thus found a solution. 

The magic keyword is "merge-base". It is the common ancestor used by `git merge`. The merge-base is automatically computed by git using a "most recent common ancestor" algorithm. The merge-base can not be overwritten. BUT! we can create a fake history using `git replace --graft` such that `git merge` does not "see" the "Custom merge commit" that gitJournal created and thus it chooses the merge-base we want it to!

simplified steps of the new sync script:
1) create a branch "branch_last_merge" which indicates where the last sync has been made if it does not exist.
2) create a copy Y of the commit X which "git_journal_branch" points to but with "branch_last_merge" as parent and create a reference which redirects all access to X to Y. This can be done with a single command in git (see references 2) and 3)): 
```bash
git replace --graft --force "$git_journal_branch" "$branch_last_merge"
```

3) merge "git_journal_branch" into main. (this now uses Y instead of X)
4) move "branch_last_merge" forward to main

References which helped me:

1) [not the answer I was looking for but interesting](https://stackoverflow.com/a/63987236)
2) [I can choose the merge-base using grafts](https://stackoverflow.com/a/32359327)
3) [I should use replace instead of grafts](https://stackoverflow.com/a/6802005)
4) [Git - git-replace Documentation](https://git-scm.com/docs/git-replace)
5) [How can I tell if one commit is a descendant of another commit?](https://stackoverflow.com/a/13526591)

And here is my updated merge script:

```bash
#!/usr/bin/env bash

# print executed commands
set -o xtrace

# fail fast
set -Eeuo pipefail

branch='git-journal-shift'
main_branch='main'

# branch that points to commit where last merge was performed (this is used as the merge-base!)
branch_last_merge='last-merge'

# temporary branch which indicates that a graft ref for that commit should automatically be removed on next script execution.
temp_todo_remove_graft='todo_remove_graft'

# auto commit main first
scripts/auto_commit.sh

# abort if merge is ongoing
MERGE_INDICATOR_FILE="$(git rev-parse --git-dir)"/MERGE_HEAD
if test -f "$MERGE_INDICATOR_FILE"; then
    echo "*********************** merge is ongoing, commit first. ***********************"
    exit 1;
fi

# create last-merge branch if not exists
if ! git rev-parse --verify "$branch_last_merge" ; then
  git branch --force "$branch_last_merge" "$main_branch"
fi

# fetch branch
git fetch origin $branch:$branch

# merge if branch is not ancestor of main branch
if ! git merge-base --is-ancestor "$branch" "$main_branch" ; then
  git branch --force "$temp_todo_remove_graft" "$branch"
  git replace --graft --force "$branch" "$branch_last_merge"
  git merge --no-ff "$branch"
fi

# clean up: remove graft ref and temporary branch (ignore errors)
if git rev-parse --verify "$temp_todo_remove_graft" ; then
  git replace -d "$temp_todo_remove_graft" || true
  git branch -D "$temp_todo_remove_graft" || true
fi

# fast forward branch to main (the code above ensures "branch" is an ancestor of main)
git branch --force "$branch" "$main_branch"

# push
git push origin "$branch"
git push origin "$main_branch"

# update branch which points to commit at last merge
git branch --force "$branch_last_merge" "$main_branch"

# 
echo "*********************** ALL GOOD! **************************"
```

And the auto commit script which is executed by the merge script: `scripts/auto_commit.sh`

```bash
#!/usr/bin/env bash

# print executed commands
set -o xtrace

# fail fast
set -Eeuo pipefail

commit_message='m'

# abort if merge is ongoing
MERGE_INDICATOR_FILE="$(git rev-parse --git-dir)"/MERGE_HEAD
if test -f "$MERGE_INDICATOR_FILE"; then
    echo "*********************** merge is ongoing, commit first. ***********************"
    exit 1;
fi

# ref: https://stackoverflow.com/a/2658301
function evil_git_num_untracked_files {
  expr `git status --porcelain 2>/dev/null| grep "^??" | wc -l` 
}

# commit & pull & push
# git commit returns exit status 1 if nothing to commit so need to check first
if (( $(evil_git_num_untracked_files) > 0 )); then
  # there are untracked files
  git add -A
  git commit -a -m 'm'
fi
# if HEAD is dirty commit
git diff-index --quiet HEAD || git commit -a -m "$commit_message"
git pull
git push
```
